Description: Tolerate GTK3 deprecation warnings
Author: Mike Gabriel <mike.gabriel@das-netzwerkteam.de>

--- a/configure.ac
+++ b/configure.ac
@@ -155,7 +155,7 @@
 AM_PATH_GTK_3_0
 
 dnl CFLAGS
-CFLAGS="$CFLAGS -Wall -Werror -Wempty-body -Wformat-security -Winit-self -Warray-bounds -Wimplicit-function-declaration -lm"
+CFLAGS="$CFLAGS -Wall -Wempty-body -Wformat-security -Winit-self -Warray-bounds -Wimplicit-function-declaration -lm"
 
 AC_SUBST(BAMFDAEMON_CFLAGS)
 AC_SUBST(BAMFDAEMON_LIBS)
